source-code/core/src/lint/RFC.md

```diff
@@ -0,0 +1,128 @@
+
+2. We need to extend the Resource type on each level (`Resource`, `Message`, `Pattern`) because each level could have lint information attached to it.
+
+	- "missing key" would be on the `Resource` level
+	- "missing parameter" would be on the `Message` level
+	- "unknown parameter" would be on the `Pattern` level
+
+	Maybe `Pattern` requires another level in the future, but if we solve the problem for the first 3 levels, adding the 4th should be easy.
+
+	Kind of similar to the `metadata` property which can be present on all levels. If we solve the problem for `lint`, we can get rid of the non-typed `metadata` property and plugin authors can use a slightly different approach (additional properties should be scoped to a "namespace"-like object). The result would be a `Resource` with correctly typed `metadata` property.
+
+	See [`experiment-2.ts`](https://github.com/ivanhofer/inlang/blob/lint/source-code/core/src/lint/experiment-2.ts) for an example how this could be done. The only downside is, that those generics make the code harder to read. And VS Code will show jibberish when hovering over a variable (see screenshots below). But the only alternative I can see is to define each Permutation manually which is even worse.
+
+3. What information would a lint error emit?
+
+	```ts
+	type LintInformation = {
+		type: LintRule // which rule emitted the error
+		message: string // a string that informs what is wrong
+	}
+	```
+	Question: Anything else?
+
+	Question: Should any lint rule be able to extend this object?
+
+<!------------------------------------------------------------------------------------------------>
+
+## Configure Lint behavior
+
+`inlang` will provide some lint functionality that can be configured or disabled.
+It does not need to be as complicated as eslint, but it should follow the same design when it comes to the `rules`.
+The `inlang.config.js` will be extended with a new `linting` property.
+
+```js
+/**
+ * @type {import("@inlang/core/config").DefineConfig}
+ */
+export async function defineConfig(env) {
+  	return {
+   	referenceLanguage: 'en,
+   	languages: ['en', 'de'],
+		linting: {
+			missing_key: false, // disable this specific lint rule
+			missing_key: 'warn', // override the level
+			missing_key: { 'error', { ignore: ['test-key'] }}, // configure the linting behavior where needed
+		}
+  	}
+}
+```
+
+See [`experiment-3.ts`](https://github.com/ivanhofer/inlang/blob/lint/source-code/core/src/lint/experiment-3.ts) for more details.
+
+Because we are in control of all lint rules, we can have strong type support for all the lint configurations.
+
+### extending lint rules
+
+How do we want to deal with this?
+
+Question: Should anyone be able to write and add custom lint rules? _Probably yes._
```

Comment:
Yes, I was also thinking of other useful non-technical extensions e.g. automatic grammar checks etc.

I didn't mention it anywhere but this means that the whole linting process needs to be `async` because a plugin might make some api calls.

Reply:
> I didn't mention it anywhere but this means that the whole linting process needs to be async because a plugin might make some api calls. 

That is true but potentially also difficult to design? If every rule can be async, we need to cache the lint results. I understand why you opened https://github.com/inlang/inlang/discussions/354

Reply:
I don't think that it is that difficult to design.